### PR TITLE
Add ability to mock kernel feature prober and expand BPF map tests

### PR DESCRIPTION
**Pull request by christarazi**
See commit msgs.

Followup from https://github.com/cilium/cilium/pull/14853

### CODE REVIEW

**Comment by christarazi** [comment]
test-me-please

File: pkg/datapath/linux/probes/probes.go [hunk 1]
```diff
@@ -151,6 +151,14 @@ type Features struct {
...
 }
 
+// Prober abstracts the notion of a kernel feature prober. This is useful for
+// testing purposes as it allows us to mock out the kernel, enabling control
+// control over what features are returned.
+type Prober interface {
+	// Probe returns the kernel feaures available on machine.
+	Probe() Features
+}
+
```
Comment (joestringer) [inline pkg/datapath/linux/probes/probes.go]: Usually we define the interface at the place where it will be consumed, I think this is in `pkg/bpf` instead since that's where the function is that receives this type. Otherwise we are forced to pull in this package with the concrete implementation at the same time that we pull in the interface. Is there a motivation for defining this interface in the probes package?

File: pkg/bpf/map_linux_test.go [hunk 2]
```diff
@@ -567,6 +568,57 @@ func (s *BPFPrivilegedTestSuite) TestCheckAndUpgrade(c *C) {
...
+					int(unsafe.Sizeof(TestValue{})),
+					maxEntries,
+					0,
+					0,
+					ConvertKeyValue).WithCache()
+				_, err := upgradeMap.OpenOrCreate()
+				c.Assert(err, IsNil)
+
+				upgrade := upgradeMap.CheckAndUpgrade(&upgradeMap.MapInfo)
+				c.Assert(upgrade, Equals, false)
```
Comment (joestringer) [inline pkg/bpf/map_linux_test.go:619]: If I follow right, the `upgrade == false` case here is basically saying that:
* Typically, LRU requires pre-allocation.
* Given the underlying lack of LRU support from the "kernel" prober, this map type would actually use hashmap
* Since the map type is switched to hashmap, now pre-allocation *can* be disabled
* When we try to upgrade the map, defining that its type should be LRU, there's no intermediate state where we decide that the map *should* be upgraded because the desired type is LRU (or the preallocation flags are mismatched)
* Instead, every single time the map info is evaluated, the type & flags are evaluated first and then the upgrade decision is made based on the attributes afterwards.

At a glance I found this a little bit cryptic in the reading of the test, you might consider stashing a comment something like this in the code. But not such a big deal.

**Review by joestringer — APPROVED** [review]
Couple of minor comments, otherwise LGTM!

File: pkg/datapath/linux/probes/probes.go (hunk 1 above)
Comment (christarazi) [reply pkg/datapath/linux/probes/probes.go]: Good point. I initially thought to provide the interface inside the `probes` package simply because proximity and that `pkg/bpf` was already importing `probes` anyway. However, that doesn't preclude the pattern of providing the interface at the consumer site, so I've made that change.

File: pkg/bpf/map_linux_test.go (hunk 2 above)
Comment (christarazi) [reply pkg/bpf/map_linux_test.go:619]: Thanks, I'll add this snippet as a comment. It can indeed be hard to follow at first glance, as it's very subtle. Also amended the commit to include a reference to the commit that fixed this to establish a link between this regression test and the fix.

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

File: pkg/bpf/map.go [hunk 3]
```diff
@@ -167,6 +173,22 @@ func GetMapType(t MapType) MapType {
 	return t
 }
 
+// SetMapTypesFromProber initializes the supported map types from the given
+// prober. This function is useful for testing purposes, as we require
+// injecting our own mocked prober.
+func SetMapTypesFromProber(prober prober) {
```
Comment (aditighag) [inline pkg/bpf/map.go]: Why is this method exported if it's only being used from the test file which is in the same package?

**Review by aditighag — CHANGES_REQUESTED** [review]
Cool 🚀

File: pkg/bpf/map.go (hunk 3 above)
Comment (christarazi) [reply pkg/bpf/map.go]: I figured it might be useful in the future, but I can unexport it for now. If it needs to be exported at some point, we can make that change.

**Review by aditighag — APPROVED** [review]
